proposals/025-lambda-1.md

```diff
@@ -0,0 +1,224 @@
+# SP #025: Lambda Expressions (Immutable Capture)
+
+This proposal adds initial support for lambda expressions in Slang.
+The initial proposal is to support immutable capture of variables in the surrounding scope.
+This means that the lambda can only read the values of the captured variables, not modify them.
+
+## Status
+
+Status: In Implementation
+
+Implementation: [PR 6914](https://github.com/shader-slang/slang/pull/6914)
+
+Author: Yong He
+
+Reviewer: Theresa Foley
+
+## Background
+
+SP009 introduced `IFunc` interface to represent callable objects. This allowed Slang code to
+pass around functions as first-class values by defining types that implement `IFunc`.
+However, this approach is not very convenient for users, as it requires defining a new type for each function
+that needs to be passed around.
+
+This problem can be solved with lambda expressions, which enables the compiler to synthesize such
+boilerplate types automatically. The recent cooperative matrix 2 SPIRV extension introduced several opcodes
+such as Reduce, PerElement, Decode etc. that can expressed naturally with lambda expressions.
+
+## Proposal
+
+The proposal is to add the following syntax for lambda expressions:
```

Comment:
Are all accessible values captured? Is there any way to write a "pure" lambda?

Reply:
Values are captured on-demand. Unreferenced variables are not captured. To write a "pure" lambda, just make sure the lambda doesn't reference any local variables or parameters in the outer function.